### PR TITLE
Eve: Add basic pattern matching capabilities

### PR DESCRIPTION
**Pull request by tehrengruber**
While testing the PAST nodes I became tired of always writing out the entire nodes and the resulting tests were often testing more then necessary. This PR introduces basic pattern matching facilities that allow comparing two values and showing the differences between the two. Note that no capturing is supported, but this could be added at a later point.

Example:
```python
import foo_ir
foo_ir_ = ModuleWrapper(foo_ir)
assert foo_ir_.Foo(bar="baz").matches(foo_ir.Foo(bar="baz", foo="bar"))
assert not foo_ir_.Foo(bar="bar").matches(foo_ir.Foo(bar="baz", foo="bar"))
```

### CODE REVIEW

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:34]: It doesn't make sense to iterate the generators with all the errors if they haven't been requested by the user. We could just return `False` as soon as the first error is found:

```suggestion
        if not raise_exception:
            return next(get_differences(self, other), None) is None
        else:
            diffs = [*get_differences(self, other)]
            if len(diffs) == 0:
                return True
            else:
                diffs_str = "\n  ".join([f"  {self.cls.__name__}{path}: {msg}" for path, msg in diffs])
                raise ValueError(f"Object and pattern don't match:\n  {diffs_str}")
```

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:107]: Please, make lines shorter

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:139]: I find this interface confusing and error-prone. I would rather use a more explicit pattern creator syntax, for example using __class_getitem__ to get something like `Pattern[SomeClass](foo="foo", bar="bar")` or `Matcher[SomeClass](foo="foo", bar="bar")`

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:63]: The pythonic way would be:
```suggestion
                yield from get_differences(a.attrs[k], getattr(b, k), path=f"{path}.{k}")
```

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:38]:
```suggestion
        fields_str = ", ".join([f"{str(k)}={str(v)}" for k, v in self.fields.items()])
        return f"{self.cls.__name__}({fields_str})"
```

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:74]:
```suggestion
        for i, (el_a, el_b) in enumerate(zip(a, b)):
            yield from get_differences(el_a, el_b, path=f"{path}[{i}]")
```
It might be even possible to make it a one-liner using `itertools.chain.from_iterable()` but I think it would harder to read.

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:21]: Missing docstring.

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:100]: I would protect the `path` argument since it's an internal argument not meant to be used in the public interface of the function.

```suggestion
def get_differences(a: Any, b: Any, *, __path: str = "") -> Iterator[tuple[str, str]]:
```

File: tests/eve_tests/unit_tests/test_pattern_matching.py
Comment (egparedes) [inline tests/eve_tests/unit_tests/test_pattern_matching.py:21]: This is more convoluted than needed and not idiomatic (the TypedDict definition is also wrong: should not contain `name`). The simple way to implement the same workflow using regular `pytest` style would be something like:

```
# Each sample contains: (case_name, a, b, expected_diff)
test_data: list[tuple[str, Any, Any, list[tuple[str, str]]]] = [
    ("int_equal", 1, 1, [])
]

@pytest.mark.parametrize("name,a, b, expected_diff", test_data)
def test_all(name,a, b, expected_diff):
    diff = list(get_differences(a, b, path="a"))
    assert diff == expected_diff
```

Additionally, I miss some test cases dealing with nested data dict/lists/objects

**Review by egparedes — CHANGES_REQUESTED** [review]
I'm not fully convinced of the scalability/composability of using `get_differences()` as main dispatcher, but it could be good enough for now. I left some suggestions to enhance the current implementation.

File: src/eve/pattern_matching.py
Comment (tehrengruber) [reply src/eve/pattern_matching.py:46]: I deliberately created this class instead of using `functools.partial` to have a good error message and to have access to `cls` during debugging when one accidentally tries to use the `ObjectPatternConstructor` instead of the `ObjectPattern` itself.

File: src/eve/pattern_matching.py
Comment (tehrengruber) [reply src/eve/pattern_matching.py:23]: I'm not too much into using `fields` as what is stored here applies to general objects for which the common name is attribute (correct me if I'm wrong here). How about `attributes`?

File: src/eve/pattern_matching.py
Comment (tehrengruber) [reply src/eve/pattern_matching.py:34]: True, I've implemented a slightly different, but similar version to what you proposed.

File: src/eve/pattern_matching.py
Comment (tehrengruber) [reply src/eve/pattern_matching.py:38]: See comment at the top.

File: src/eve/pattern_matching.py
Comment (tehrengruber) [reply src/eve/pattern_matching.py:100]: It is explicitly meant as a public argument. This is actually used in one of the tests:
```
@pytest.mark.parametrize("name,test_data", test_data)
def test_all(name, test_data):
    a, b, expected_diff = test_data["a"], test_data["b"], test_data["expected_diff"]
    diff = list(get_differences(a, b, path="a"))
```
The resulting messages contain valid python expressions to access the respective value and are very handy to be used inside the debugger.

File: tests/eve_tests/unit_tests/test_pattern_matching.py
Comment (tehrengruber) [reply tests/eve_tests/unit_tests/test_pattern_matching.py]: True, I've simplified as proposed and added more tests for nested things. Actually nesting was already tested before as the test for equality is already a recursive call.

File: src/eve/pattern_matching.py
Comment (egparedes) [reply src/eve/pattern_matching.py:37]: I proposed `fields` because it is now a widespread way to call plain data attributes of value classes (both  `dataclasses` and `pydantic` models use this convention), but `attributes` is also fine for me. The main point is to not to use `attrs`

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:39]: I really think that `raise_exception` should be a keyword-only argument here.

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:83]: Please make docstrings lines always around 80 chars, not longer.

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:84]: Types are singletons, so this could be a better way to check:

```suggestion
    if type(a) is not type(b):
```

File: src/eve/pattern_matching.py
Comment (egparedes) [reply src/eve/pattern_matching.py:114]: Why not using the suggested `yield from` ?

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:33]: Slightly more realistic example:
```suggestion
    >>> class Foo:
    ...    def __init__(self, bar, baz):
    ...        self.bar = bar
    ...        self.baz = baz
    >>> assert ObjectPattern(Foo, {"bar": 1}).match(Foo(1, 2))
```

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:159]: As I said in my previous review, I think this util causes more harm than good since makes the creation of pattern objects look exactly the same as instantiation actual objects. I think a more explicit approach should be used here

File: src/eve/pattern_matching.py
Comment (egparedes) [inline src/eve/pattern_matching.py:40]:
```suggestion
        """Return ``True`` if object pattern matches ``other`` using :func:`get_differences`.
```

**Review by egparedes — CHANGES_REQUESTED** [review]
Thanks for the changes. Except for the ModuleWrapper issue, It looks almost ready to go

File: src/eve/pattern_matching.py
Comment (tehrengruber) [reply src/eve/pattern_matching.py:159]: I missed 3 comments from your last review including this one. As discussed I see your point, but still like this solution better. I'll let this rest for a couple of days to see what I think and ask Rico and then we can revisit. Until then there is a comment and we should merge if the rest is sorted out.

File: src/eve/pattern_matching.py
Comment (tehrengruber) [reply src/eve/pattern_matching.py:40]: Mh, not a fan of this, but I've changed it. If I would be nit picking what you are proposing here is not as precise because it is unspecified what happens if the pattern does not match.

File: src/eve/pattern_matching.py
Comment (DropD) [inline src/eve/pattern_matching.py:147]:
```suggestion
    """Small wrapper to conveniently create `ObjectPattern`s for classes of a module.
```

One single convention is still better than mixing, even if it is the ugliest convention.

**Review by egparedes — COMMENTED** [review]
Minor style issues and the `ModuleWrapper` issue, which in my opinion is wrong and needs to be changed. I would only agree to merge this PR in its current state to unblock the merge of #368 if that PR is ready to go before this one.

File: src/eve/pattern_matching.py
Comment (egparedes) [reply src/eve/pattern_matching.py]: It's a minor thing. You can change it back if you prefer, but please use proper restructured text markups.

**Review by egparedes — APPROVED** [review]
It looks good to me.

**Pull request merged**
